OrbitBase/ThreadPool.cpp

```diff
@@ -8,43 +8,53 @@
 #include <thread>
 
 #include "OrbitBase/Logging.h"
+#include "absl/container/flat_hash_map.h"
 #include "absl/synchronization/mutex.h"
 
 namespace {
 
 class ThreadPoolImpl : public ThreadPool {
  public:
   explicit ThreadPoolImpl(size_t thread_pool_min_size,
-                          size_t thread_pool_max_size);
+                          size_t thread_pool_max_size,
+                          absl::Duration thread_ttl);
 
+  size_t GetPoolSize() override;
   void Schedule(std::unique_ptr<Action> action) override;
   void Shutdown() override;
   void Wait() override;
 
  private:
   bool ActionsAvailableOrShutdownInitiated();
-  // Blocking call - returns nullptr if shutdown initiated.
-  // This also decrements idle_threads_ counter (in should be
-  // done in the same scope as poping action from the queue).
+  // Blocking call - returns nullptr if the worker thread needs to exit.
   std::unique_ptr<Action> TakeAction();
+  void CleanupFinishedThreads();
   void CreateWorker();
   void WorkerFunction();
 
   absl::Mutex mutex_;
   std::list<std::unique_ptr<Action>> scheduled_actions_;
-  std::vector<std::thread> worker_threads_;
+  absl::flat_hash_map<std::thread::id, std::thread> worker_threads_;
+  std::vector<std::thread> finished_threads_;
+  size_t thread_pool_min_size_;
   size_t thread_pool_max_size_;
+  absl::Duration thread_ttl_;
   size_t idle_threads_;
   bool shutdown_initiated_;
 };
 
 ThreadPoolImpl::ThreadPoolImpl(size_t thread_pool_min_size,
-                               size_t thread_pool_max_size)
-    : thread_pool_max_size_(thread_pool_max_size),
+                               size_t thread_pool_max_size,
+                               absl::Duration thread_ttl)
+    : thread_pool_min_size_(thread_pool_min_size),
+      thread_pool_max_size_(thread_pool_max_size),
+      thread_ttl_(thread_ttl),
       idle_threads_(0),
       shutdown_initiated_(false) {
   CHECK(thread_pool_min_size > 0);
   CHECK(thread_pool_max_size >= thread_pool_min_size);
+  // Ttl should not be too small
+  CHECK(thread_ttl / absl::Nanoseconds(1) >= 1000);
```

Comment:
Is a clearer `thread_ttl >= absl::Nanoseconds(1000)` supported?

Reply:
Duration does not have operator<> == and alike.